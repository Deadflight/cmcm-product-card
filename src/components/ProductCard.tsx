import React, { createContext, CSSProperties } from 'react';
import styles from '../styles/styles.module.css';

import { useProducts } from '../hooks';
import {
  ProductContextProps,
  Product,
  onChangeArgs,
  InitialValuesProps,
} from '../interfaces';
// import { ProductTitle, ProductImage, ProductButtons } from "./";
import { ProductCardHandlers } from '../interfaces/interfaces';

export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;
export interface ProductCardProps {
  product: Product;
  children: (args: ProductCardHandlers) => JSX.Element;
  //children?: ReactElement | ReactElement[];
  className?: string;
  style?: CSSProperties;
  value?: number;
  initialValues?: InitialValuesProps;
  onChange?: (args: onChangeArgs) => void;
}

export const ProductCard = ({
  children,
  product,
  className,
  value,
  style,
  initialValues,
  onChange,
}: ProductCardProps) => {
  const {
    counter,
    maxCount,
    isMaxCountReached,
    reset,
    increaseBy,
  } = useProducts({
    onChange,
    product,
    value,
    initialValues,
  });
  return (
    <Provider
      value={{
        counter,
        product,
        maxCount,
        increaseBy,
      }}
    >
      <div className={`${styles.productCard} + ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,

          increaseBy,
          reset,
        })}
        {/* <ProductImage img={product.img} />
			<ProductTitle title={product.title} />
			<ProductButtons counter={counter} increaseBy={increaseBy} /> */}
      </div>
    </Provider>
  );
};
