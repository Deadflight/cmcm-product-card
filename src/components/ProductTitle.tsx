import { ProductContext } from './ProductCard';
import React, { useContext, CSSProperties } from 'react';
import styles from '../styles/styles.module.css';

export interface ProducTitleProps {
  title?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, className, style }: ProducTitleProps) => {
  const { product } = useContext(ProductContext);

  return (
    <span
      className={`${styles.productDescription} + ${className}`}
      style={style}
    >
      {title ? title : product.title}
    </span>
  );
};
