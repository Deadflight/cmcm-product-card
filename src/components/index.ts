import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductButtons } from './ProductButtons';
import { ProductCardHOCProps } from '../interfaces';

export { ProductCardProps } from './ProductCard';
export * from './ProductImage';
export * from './ProductTitle';
export * from './ProductButtons';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});

export default ProductCard;
