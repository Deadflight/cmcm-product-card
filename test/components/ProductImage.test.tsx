import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('should render the component with a custom image', () => {
    const wrapper = renderer.create(<ProductImage img="https:\\hola.jpg" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should show the component with the product image', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
