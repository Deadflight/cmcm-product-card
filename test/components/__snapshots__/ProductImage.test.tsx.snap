// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should render the component with a custom image 1`] = `
<img
  alt="Product Img"
  className="productImg + undefined"
  src="https:\\\\\\\\hola.jpg"
/>
`;

exports[`ProductImage should show the component with the product image 1`] = `
<div
  className="productCard + undefined"
>
  <img
    alt="Product Img"
    className="productImg + undefined"
    src="./coffee-mug.png"
  />
</div>
`;
